### PR TITLE
Add support for RegexSet

### PR DESCRIPTION
**Pull request by c-nixon**
This commit adds support for RegexSet now that it's in regex-capi

### CODE REVIEW

**Comment by c-nixon** [comment]
Regex tag in .travis.yml updated, Tests are now failing in the same way as previous pull requests!

**Comment by davidblewett** [comment]
I apologize; I did not see this PR until just now. Evidently forgot to subscribe to my own project.

**Comment by davidblewett** [comment]
It appears the Travis setup needs some love. I ran the tests on CPython 2.7.13 and CPython 3.6.1 against regex 0.2.2 and regex-capi 0.2, and they all passed.

File: rure/lib.py [hunk 1]
```diff
@@ -250,3 +250,88 @@ def shortest_match(self, haystack, start=0):
...
+
+        """ Compiles a regular expression. Once compiled, it can be used
+        repeatedly to search, split or replace text in a string.
+
+        :param res:     List of Bytestring expressions to compile
+        :param flags:   Bitmask of flags
+        :param kwargs:  Config options to pass (size_limit, dfa_size_limit)
+        """
+
+        if not all(isinstance(re, bytes) for re in res):
```
Comment (davidblewett) [inline rure/lib.py:276]: This should probably check to ensure `res` is not a string object before iterating. For instance, on Python 2.7:
```python
>>> all([isinstance(i, bytes) for i in b'foobar'])
True
```

Python 3 behaves differently here, as iterating over a byte string returns the integer representing the Unicode code point of that character:

```python
>>> all([isinstance(i, bytes) for i in b'foobar'])
False
```

**Review by davidblewett — CHANGES_REQUESTED** [review]

**Comment by davidblewett** [comment]
@c-nixon: what do you think of this diff? It simplifies the calling method so use star args, so that you just pass multiple positional args representing each expression to match. The downside of this is that on Python 2, we have to obfuscate the `flags` parameter by including it in the `options` keyword args.

```diff
diff --git a/rure/lib.py b/rure/lib.py
index 7f4bf19..b7fad64 100644
--- a/rure/lib.py
+++ b/rure/lib.py
@@ -263,16 +263,18 @@ class RureSet(object):
     match. Indeed, this is the key difference between regex sets and a single
     Regex with many alternates, since only one alternate can match at a time.
     """
-    def __init__(self, res, flags=DEFAULT_FLAGS, **options):
+    def __init__(self, *res, **options):
 
         """ Compiles a regular expression. Once compiled, it can be used
         repeatedly to search, split or replace text in a string.
 
         :param res:     List of Bytestring expressions to compile
-        :param flags:   Bitmask of flags
-        :param kwargs:  Config options to pass (size_limit, dfa_size_limit)
+        :param kwargs:  Config options to pass (flags bitmask,
+                                                size_limit,
+                                                dfa_size_limit)
         """
 
+        flags = options.pop('flags', DEFAULT_FLAGS)
         if not all(isinstance(re, bytes) for re in res):
             raise TypeError("'rure.lib.RureSet' must be instantiated with a "
                             "list of bytestrings as first argument.")
diff --git a/rure/tests/test_rureset.py b/rure/tests/test_rureset.py
index ff1783b..849e2a2 100644
--- a/rure/tests/test_rureset.py
+++ b/rure/tests/test_rureset.py
@@ -6,19 +6,25 @@ import unittest
 from rure.exceptions import CompiledTooBigError, RegexSyntaxError
 from rure.lib import CASEI, RureSet
 
+
 class TestRureSet(unittest.TestCase):
+
+    def test_type_check(self):
+        with self.assertRaises(TypeError):
+            RureSet(b"baz", u"bar", b"foo")
+
     def test_is_match(self):
         haystack = b"snowman: \xE2\x98\x83"
-        res = RureSet([b"\\p{So}"])
+        res = RureSet(b"\\p{So}")
         self.assertIsNotNone(res.is_match(haystack))
 
     def test_matches(self):
         haystack = b"foobar"
-        res = RureSet([b"baz", b"bar", b"foo"])
+        res = RureSet(b"baz", b"bar", b"foo")
         self.assertEqual(res.matches(haystack), [False, True, True])
 
     def test_set_len(self):
-        res = RureSet([b"baz", b"bar", b"foo"])
+        res = RureSet(b"baz", b"bar", b"foo")
         self.assertEqual(len(res), 3)
 
     def test_flags(self):
@@ -29,19 +35,19 @@ class TestRureSet(unittest.TestCase):
         arbitrary possibly invalid UTF-8 bytes, such as \xFF.
         (When Unicode mode is enabled, \xFF won't match .)
         """
-        pattern = [b"."]
+        pattern = b"."
         haystack = b"\xFF"
         res = RureSet(pattern, flags=CASEI)
         self.assertTrue(res.is_match(haystack))
 
     def test_compile_error(self):
         try:
-            RureSet([b"("])
+            RureSet(b"(")
         except RegexSyntaxError as err:
             self.assertIn("Unclosed parenthesis", err.message)
 
     def test_compile_error_size_limit(self):
         try:
-            RureSet([b"\\w{100}"], size_limit=0)
+            RureSet(b"\\w{100}", size_limit=0)
         except CompiledTooBigError as err:
             self.assertIn("exceeds size", err.message)
```

**Comment by ghost** [comment]
Hey @davidblewett @c-nixon - Thanks for this feature! A while back, before PyO3 stabilised, I tried writing RegexSet bindings for Python but it "failed" (Performance was, for some reason, unacceptably terible). I'm really happy to see this. This could be really, really helpful for some projects I've worked on using mass-regex-matching. :)

Small request: Could a version with this feature be pushed to PyPI?

**Comment by davidblewett** [comment]
@cathalgarvey I haven't had time to work through the packaging problems since my last attempt to push an update. I had planned on vendoring the upstream Rure C API and updating the build procedure to use TravisCI to build Mac and manylinux wheels. I just haven't had the time. I would appreciate assistance, if you are familiar with Python packaging and building binary wheels.

**Comment by ghost** [comment]
Sadly I'm not, I got stuck previously pushing a pure-rust binding before PyO3 stabilised, and that was probably a much "easier" task.

Thanks all the same for this work: I can pip install from git for now until things stabilise.

On 1 April 2019 01:01:37 GMT+01:00, David Blewett <notifications@github.com> wrote:
>@cathalgarvey I haven't had time to work through the packaging problems
>since my last attempt to push an update. I had planned on vendoring the
>upstream Rure C API and updating the build procedure to use TravisCI to
>build Mac and manylinux wheels. I just haven't had the time. I would
>appreciate assistance, if you are familiar with Python packaging and
>building binary wheels.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/davidblewett/rure-python/pull/10#issuecomment-478398043

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by davidblewett** [comment]
@cathalgarvey I pushed up binary wheels for MacOS Mojave (10.14) as well as manylinux for version 0.2.1. Can you test to see if they work for you? The test suite passes on MacOS using Python 3.7 from Homebrew.

**Pull request merged**